reasoner/planner/RecursivePlanner.java

```diff
@@ -0,0 +1,383 @@
+            initialiseDependencies(conjunctionNode, resolvable, resolvableBounds);
+            if (!resolvable.isNegated()) currentBounds.addAll(consideredVariables(resolvable));
+        }
+    }
+
+    private void initialiseDependencies(ConjunctionNode conjunctionNode, Resolvable<?> resolvable, Set<Variable> resolvableBounds) {
+        if (resolvable.isConcludable()) {
+            Set<ResolvableConjunction> cyclicDependencies = conjunctionNode.cyclicDependencies(resolvable.asConcludable());
+            for (CallMode callMode : triggeredCalls(resolvable.asConcludable(), resolvableBounds, Optional.empty())) {
+                recursivelyGenerateOrderingChoices(callMode);
+                if (!cyclicDependencies.contains(callMode.conjunction)) {
+                    plan(callMode); // Acyclic dependencies can be fully planned
+                }
+            }
+        } else if (resolvable.isNegated()) {
+            iterate(resolvable.asNegated().disjunction().conjunctions()).forEachRemaining(conjunction -> {
+                CallMode callMode = new CallMode(conjunction, resolvableBounds);
+                recursivelyGenerateOrderingChoices(callMode);
+                plan(callMode);
+            });
+        }
+    }
+
+    private void recursivelyGenerateOrderingChoices(CallMode callMode) {
+        if (!candidateOrderings.containsKey(callMode)) {
+            candidateOrderings.put(callMode, null); // Guard
+            ConjunctionNode conjunctionNode = conjunctionGraph.conjunctionNode(callMode.conjunction);
+            answerCountEstimator.buildConjunctionModel(callMode.conjunction);
+
+            // Two orderings for the same CallMode with identical cyclic-concludable modes are interchangeable in the component-plan
+            //      -> We only need to keep the cheaper one.
+            Map<Set<Pair<Concludable, Set<Variable>>>, OrderingChoice> bestChoice = new HashMap<>();
+            PartialOrderReductionSearch porSearch = new PartialOrderReductionSearch(logicMgr.compile(callMode.conjunction), callMode.bounds);
+            for (List<Resolvable<?>> ordering : porSearch.allOrderings()) {
+                initialiseOrderingDependencies(conjunctionNode, ordering, callMode.bounds);
+                OrderingChoice orderingChoice = summariseOrdering(conjunctionNode, ordering, callMode.bounds);
+                if (!bestChoice.containsKey(orderingChoice.cyclicConcludableBounds) ||
+                        bestChoice.get(orderingChoice.cyclicConcludableBounds).acyclicCost > orderingChoice.acyclicCost) {
+                    bestChoice.put(orderingChoice.cyclicConcludableBounds, orderingChoice);
+                }
+            }
+            candidateOrderings.put(callMode, new HashSet<>(bestChoice.values()));
+        }
+    }
+
+    private OrderingChoice summariseOrdering(ConjunctionNode conjunctionNode, List<Resolvable<?>> ordering, Set<Variable> inputBounds) {
```

Comment:
everywhere you use `bounds`, should we be using `mode`?

Reply:
Good point